In [10]:
import pandas as pd
pd.set_option("display.max_columns", None)

In [3]:
df =pd.read_csv("../Data/Raw/Output/01_data_trabajo.csv")

In [5]:
print(df.shape)
print(df["movieId"].nunique())
print(df["userId"].nunique())
print(df["rating"].describe())
print(df["title"].nunique())

(70000, 32)
3546
755
count    70000.000000
mean         3.606986
std          1.045944
min          0.500000
25%          3.000000
50%          4.000000
75%          4.500000
max          5.000000
Name: rating, dtype: float64
3544


Entendimiento de los datos

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   userId                70000 non-null  int64  
 1   movieId               70000 non-null  int64  
 2   rating                70000 non-null  float64
 3   timestamp             70000 non-null  int64  
 4   title                 70000 non-null  str    
 5   genres_x              70000 non-null  str    
 6   imdbId                70000 non-null  int64  
 7   tmdbId                70000 non-null  float64
 8   budget                70000 non-null  int64  
 9   genres_y              70000 non-null  str    
 10  homepage              28911 non-null  str    
 11  id                    70000 non-null  int64  
 12  keywords              70000 non-null  str    
 13  original_language     70000 non-null  str    
 14  original_title        70000 non-null  str    
 15  overview              70000 no

In [13]:
df.sample(4)

,userId,movieId,rating,timestamp,title,genres_x,imdbId,tmdbId,budget,genres_y,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title_x,vote_average,vote_count,movie_id,title_y,cast,crew
5301,59,22,4.5,1111475159,Copycat (1995),Crime|Drama|Horror|Mystery|Thriller,112722,1710.0,0,"[{""id"": 18, ""name"": ""Drama""}, {""id"": 53, ""name...",NaN,1710,"[{""id"": 796, ""name"": ""police brutality""}, {""id...",en,Copycat,An agoraphobic psychologist and a female detec...,14.211699,"[{""name"": ""Regency Enterprises"", ""id"": 508}, {...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",1995-10-27,0,124.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,One man is copying the most notorious killers ...,Copycat,6.4,194,1710,Copycat,"[{""cast_id"": 8, ""character"": ""Helen Hudson"", ""...","[{""credit_id"": ""561cd9b4c3a368225e0034cc"", ""de..."
47565,534,2396,5.0,1042672546,Shakespeare in Love (1998),Comedy|Drama|Romance,138097,1934.0,25000000,"[{""id"": 10749, ""name"": ""Romance""}, {""id"": 36, ...",NaN,1934,"[{""id"": 497, ""name"": ""shakespeare""}, {""id"": 20...",en,Shakespeare in Love,Young Shakespeare is forced to stage his lates...,31.792964,"[{""name"": ""Miramax Films"", ""id"": 14}, {""name"":...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",1998-12-11,289317794,122.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,Love is the only inspiration.,Shakespeare in Love,6.8,796,1934,Shakespeare in Love,"[{""cast_id"": 18, ""character"": ""Will Shakespear...","[{""credit_id"": ""52fe4323c3a36847f803d3fd"", ""de..."
55838,606,2953,4.0,1496447126,Home Alone 2: Lost in New York (1992),Children|Comedy,104431,772.0,18000000,"[{""id"": 35, ""name"": ""Comedy""}, {""id"": 10751, ""...",NaN,772,"[{""id"": 65, ""name"": ""holiday""}, {""id"": 242, ""n...",en,Home Alone 2: Lost in New York,"Instead of flying to Florida with his folks, K...",81.487685,"[{""name"": ""Twentieth Century Fox Film Corporat...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",1992-11-19,358991681,120.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,He's up past his bedtime in the city that neve...,Home Alone 2: Lost in New York,6.3,2395,772,Home Alone 2: Lost in New York,"[{""cast_id"": 12, ""character"": ""Kevin"", ""credit...","[{""credit_id"": ""52fe4274c3a36847f802001b"", ""de..."
66825,716,1339,4.0,1211732532,Dracula (Bram Stoker's Dracula) (1992),Fantasy|Horror|Romance|Thriller,103874,6114.0,40000000,"[{""id"": 10749, ""name"": ""Romance""}, {""id"": 27, ...",NaN,6114,"[{""id"": 596, ""name"": ""adultery""}, {""id"": 2158,...",en,Dracula,When Dracula leaves the captive Jonathan Harke...,45.022964,"[{""name"": ""American Zoetrope"", ""id"": 70}, {""na...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",1992-11-13,215862692,128.0,"[{""iso_639_1"": ""la"", ""name"": ""Latin""}, {""iso_6...",Released,Love never dies.,Dracula,7.1,1056,6114,Dracula,"[{""cast_id"": 2, ""character"": ""Dracula"", ""credi...","[{""credit_id"": ""52fe443ec3a36847f808a953"", ""de..."


Data understanding - comprobacion de nulos y tipo de datos

In [17]:
data_dict = pd.DataFrame({
    'Columna': df.columns,
    'Tipo': df.dtypes.values,
    'Nulos': df.isnull().sum().values,
    '% Nulos': round(df.isnull().mean() * 100, 2)
})
display (data_dict)

,Columna,Tipo,Nulos,% Nulos
userId,userId,int64,0,0.00
movieId,movieId,int64,0,0.00
rating,rating,float64,0,0.00
timestamp,timestamp,int64,0,0.00
title,title,str,0,0.00
genres_x,genres_x,str,0,0.00
imdbId,imdbId,int64,0,0.00
tmdbId,tmdbId,float64,0,0.00
budget,budget,int64,0,0.00
genres_y,genres_y,str,0,0.00


De este análisis destacamos que no hay variables numericas que deban ser str y al revés. No habrá que hacer modificaciones de este tipo. 
Excepto *ralease_date* habra que modificarlo a tipo fecha al igual que *timestamp*

In [18]:
df.shape

(70000, 32)

In [20]:
pd.set_option('display.float_format','{:,.2f}'.format)
df.describe().T

,count,mean,std,min,25%,50%,75%,max
userId,"70,000.00",381.33,218.60,1.00,187.00,396.00,567.00,755.00
movieId,"70,000.00","19,843.47","33,774.81",1.00,"1,193.00","2,889.00","8,874.00","161,580.00"
rating,"70,000.00",3.61,1.05,0.50,3.00,4.00,4.50,5.00
timestamp,"70,000.00","1,218,234,406.55","233,065,218.86","828,163,431.00","993,932,420.00","1,225,302,990.50","1,447,703,970.00","1,574,253,749.00"
imdbId,"70,000.00","354,778.34","513,277.86","6,864.00","108,550.00","120,878.50","347,149.00","5,215,952.00"
tmdbId,"70,000.00","15,647.78","40,855.24",5.00,587.00,"2,074.00","10,220.00","376,659.00"
budget,"70,000.00","49,104,643.81","49,983,875.41",0.00,"13,000,000.00","31,000,000.00","70,000,000.00","380,000,000.00"
id,"70,000.00","15,647.78","40,855.24",5.00,587.00,"2,074.00","10,220.00","376,659.00"
popularity,"70,000.00",50.85,50.66,0.01,21.19,41.08,66.17,875.58
revenue,"70,000.00","241,741,529.17","279,224,161.94",0.00,"46,189,568.00","149,270,999.00","351,583,407.00","2,787,965,087.00"


Del metodo describe destacamos:
- las variables id y movie_id son iguales
- No hay valores booleanos
- El valor rating va de 0,5 a 5.
- El valor tiemstamp hay que modificarlo a fecha
- 'Budget' aparece con algunos valores 0, consideramos que un valor 0 es un valor nulo ya que debe tener asociado un budget cada pelicula


Comprobamos valores duplicados

In [24]:
df.duplicated().sum()

np.int64(0)

Comprobamos valores unicos

In [25]:
df.value_counts()

userId  movieId  rating  timestamp   title                                                      genres_x                               imdbId  tmdbId  budget    genres_y                                                                                                                                  homepage                                                id   keywords                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [ ]:
df["overview"].value_counts()
# el restulado es 3544 por lo que coincide casi en la totalidad con el numero de "movieId" unicos (3546)

overview
A man with a low IQ has accomplished great things in his life and been present during significant historic events - in each case, far exceeding what anyone imagined he could do. Yet, despite all the things he has attained, his one true love eludes him. 'Forrest Gump' is the story of a man who rose above his challenges, and who proved that determination, courage, and love are more important than ability.        368
A burger-loving hit man, his philosophical partner, a drug-addled gangster's moll and a washed-up boxer converge in this sprawling, comedic crime caper. Their adventures unfurl in three stories that ingeniously trip back and forth in time.                                                                                                                                                                               361
Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he pu

In [ ]:
df["status"].value_counts()
#la columna status no aporta valor, solo hay como dato "released"

status
Released    70000
Name: count, dtype: int64

Hemos visto que los datos estan todos en minúscula a excepcion de la primera letra de cada palabra, por lo que no hay que unificar.
Adicionalmente, no hay valores tipo object por lo que tampoco habria que unificar esto.

Modificamos a tipo fecha las variables *timestamp* y *release_date*

In [34]:
df["timestamp"] = pd.to_datetime(df["timestamp"])

In [35]:
df["timestamp"].dtype

dtype('<M8[s]')

In [36]:
df["timestamp"].head()

0   2006-05-17 15:34:04
1   2006-05-17 12:21:50
2   2006-05-17 12:20:14
3   2006-05-17 12:14:57
4   2006-05-17 12:14:39
Name: timestamp, dtype: datetime64[s]

In [37]:
df["release_date"] = pd.to_datetime(df["release_date"], errors="coerce")

In [38]:
df["release_date"].dtype

dtype('<M8[us]')

In [39]:
df["release_year"] = df["release_date"].dt.year
df["rating_year"] = df["timestamp"].dt.year

Modficiacion de las columnas *id* a tipo str

In [40]:
df["tmdbId"].head()

0   680.00
1   872.00
2   826.00
3   782.00
4   165.00
Name: tmdbId, dtype: float64

In [41]:
#cambiamos la columna "tmsbId" de tipo float a tipo int
df["tmdbId"] = df["tmdbId"].astype(int)

In [42]:
df["tmdbId"].dtype

dtype('int64')

In [45]:
id_columns = ["userId", "movieId", "imdbId", "tmdbId", "movieId", "id"]


In [47]:
for col in id_columns:
    df[col] = df[col].astype(str)

In [48]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 34 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   userId                70000 non-null  str           
 1   movieId               70000 non-null  str           
 2   rating                70000 non-null  float64       
 3   timestamp             70000 non-null  datetime64[s] 
 4   title                 70000 non-null  str           
 5   genres_x              70000 non-null  str           
 6   imdbId                70000 non-null  str           
 7   tmdbId                70000 non-null  str           
 8   budget                70000 non-null  int64         
 9   genres_y              70000 non-null  str           
 10  homepage              28911 non-null  str           
 11  id                    70000 non-null  str           
 12  keywords              70000 non-null  str           
 13  original_language     70000

In [49]:
numericas = df.select_dtypes(include=["int64", "float64"])

In [50]:
categoricas = df.select_dtypes(include=["object", "str"])

In [55]:
for col in df.select_dtypes(include="str").columns:
    print(col.upper()
          )
    print(df[col].value_counts())

USERID
userId
548    1936
187    1039
626     824
426     750
606     700
       ... 
656      10
241       8
524       8
689       8
668       6
Name: count, Length: 755, dtype: int64
MOVIEID
movieId
356      368
296      361
318      350
593      323
2571     314
        ... 
3322       1
44815      1
58839      1
70643      1
79800      1
Name: count, Length: 3546, dtype: int64
TITLE
title
Forrest Gump (1994)                 368
Pulp Fiction (1994)                 361
Shawshank Redemption, The (1994)    350
Silence of the Lambs, The (1991)    323
Matrix, The (1999)                  314
                                   ... 
3 Strikes (2000)                      1
ATL (2006)                            1
Leatherheads (2008)                   1
Bandslam (2009)                       1
Life During Wartime (2009)            1
Name: count, Length: 3544, dtype: int64
GENRES_X
genres_x
Drama                                                                            4147
Comedy              

De la distribucion por categorias vemos lo siguiente:
- coincide en cantidad el movieId, title,imdbId, tmdbId, id, original_title, title_y - por lo que las uniones se han hecho correctamente.
Forrest Gump                368
Pulp Fiction                361
The Shawshank Redemption    350
The Silence of the Lambs    323
The Matrix                  314
- el idioma original mas repetido es "English"
- El userId que más ha puntuado es el 548 y los que menos han puntuado lo han hecho una vez
- En genre_x - Drama es el genero más puntuado - coincidiendo con genre_y (esta columan no da informacion addicional)
- Las columnas: "keywords", spoken_languages, released, tagline, title_y, cast, crew, homepage - no aportan informacion relevante para el análisis.
Además de existir columnas que repiten la misma información. 
Columnas como "production_companies" deberán ser limpiadas dejando solo el "name" y "production_countries" limpiar solo el nombre

